packages/commonwealth/client/scripts/views/pages/CommunityGroupsAndMembers/Groups/common/GroupForm/Allowlist/Allowlist.tsx

```diff
@@ -0,0 +1,259 @@
+import { CWTextInput } from 'views/components/component_kit/new_designs/CWTextInput';
+import MembersSection, {
+  Member,
+} from '../../../../Members/MembersSection/MembersSection';
+import { BaseGroupFilter } from '../../../../Members/index.types';
+import { useMemberData } from '../../../../common/useMemberData';
+import { getTotalPages } from '../../helpers/index';
+import './Allowlist.scss';
+
+const tableColumns: (isStakedCommunity: boolean) => CWTableColumnInfo[] = (
+  isStakedCommunity,
+) => [
+  {
+    key: 'name',
+    header: 'Users',
+    hasCustomSortValue: true,
+    numeric: false,
+    sortable: true,
+  },
+  {
+    key: 'address',
+    header: 'Address',
+    hasCustomSortValue: true,
+    numeric: false,
+    sortable: true,
+  },
+  {
+    key: 'stakeBalance',
+    header: 'Stake',
+    hasCustomSortValue: true,
+    numeric: true,
+    sortable: true,
+    hidden: !isStakedCommunity,
+  },
+  {
+    key: 'groups',
+    header: 'Groups',
+    hasCustomSortValue: true,
+    numeric: false,
+    sortable: false,
+  },
+];
+
+type AllowlistProps = {
+  allowedAddresses: string[];
+  setAllowedAddresses: (
+    value: ((prevState: string[]) => string[]) | string[],
+  ) => void;
+};
+
+const baseFilterOptions = [
+  { type: 'header', label: 'Filters' },
+  { label: 'All community', value: 'all-community' },
+  { label: 'Allowlisted', value: 'allow-specified-addresses' },
+  { label: 'Not allowlisted', value: 'not-allow-specified-addresses' },
+  { type: 'header-divider', label: 'Groups' },
+  { label: 'All groups', value: 'in-group' },
+  { label: 'Ungrouped', value: 'not-in-group' },
+];
+
+const MEMBERS_PER_PAGE = 10;
+
+const Allowlist = ({
+  allowedAddresses,
+  setAllowedAddresses,
+}: AllowlistProps) => {
+  useUserActiveAccount();
+
+  const [searchFilters, setSearchFilters] = useState({
+    searchText: '',
+    groupFilter: 'all-community',
+  });
+
+  const [currentPage, setCurrentPage] = useState<number>(1);
+
+  const { data: memberships } = useRefreshMembershipQuery({
+    communityId: app.activeChainId(),
+    address: app?.user?.activeAccount?.address,
+    apiEnabled: !!app?.user?.activeAccount?.address,
+  });
+
+  const isStakedCommunity = !!app.config.chains.getById(app.activeChainId())
+    .namespace;
+
+  const tableState = useCWTableState({
+    columns: tableColumns(isStakedCommunity),
+    initialSortDirection: APIOrderDirection.Desc,
+  });
+
+  const debouncedSearchTerm = useDebounce<string>(
+    searchFilters.searchText,
+    500,
+  );
+
+  const memoizedAddresses = useMemo(() => {
+    return allowedAddresses;
+    // we disable the exhaustive-deps because we don't want to refresh on changed allowedAddresses because it will
+    // update the displayed list while the boxes are being checked which is a bit jarring from a UI perspective
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [searchFilters]);
```

Comment:
So this fixes:
> 1. Getting this weird behavior when trying to add a new member to allow list, during group edit flow.

The issue there is that when the checkbox is checked, the allowlists table auto-updates. Since it is on unallowlisted filter, it removes the row from the table immediately. This memoizes it so that it doesn't remove the row from the table until the search filters change (Either the dropdown filter, or the search text).

Reply:
> the table should have an internal state to have display-control over the selected elements, even on the unallowlisted filter

How do we propogate this information up to the form level to submit? Usually you would use state, but if CWTable has its internal state of allowedAddresses how do we push up the state?

> Or if we want to use the existing solution, a ref to hold allowedAddresses value which updates on searchFilters change

I think we have the same issue with this, state needs to be propogated to the parent component.

Reply:
Ok discussed offline. We will have a fast-follow ticket to fix it here:
https://github.com/hicommonwealth/commonwealth/issues/7889